libcxx/test/std/algorithms/alg.nonmodifying/alg.find/ranges.find_last_if.pass.cpp

```diff
@@ -0,0 +1,364 @@
+
+    {
+      int a[] = {4, 4, 4, 4};
+      std::same_as<std::ranges::subrange<It>> auto ret = std::ranges::find_last_if(It(a), Sent(It(a + 4)), [](int x) { return x == 4; });
+      assert(base(ret.begin()) == a + 3);
+      assert(*ret.begin() == 4);
+    }
+
+    {
+      int a[] = {4, 4, 4, 4};
+      std::same_as<std::ranges::borrowed_subrange_t<int (&)[4]>> auto ret = std::ranges::find_last_if(a, [](int x) { return x == 4; });
+      assert(base(ret.begin()) == a + 3);
+      assert(*ret.begin() == 4);
+    }
+
+  }
+
+  {// Test when the element being searched is the first one
+
+    {
+      int a[] = {4, 1, 2, 3};
+      std::same_as<std::ranges::subrange<It>> auto ret = std::ranges::find_last_if(It(a), Sent(It(a + 4)), [](int x) { return x == 4; });
+      assert(base(ret.begin()) == a);
+      assert(*ret.begin() == 4);
+    }
+
+    {
+      int a[] = {4, 1, 2, 3};
+      std::same_as<std::ranges::borrowed_subrange_t<int (&)[4]>> auto ret = std::ranges::find_last_if(a, [](int x) { return x == 4; });
+      assert(base(ret.begin()) == a);
+      assert(*ret.begin() == 4);
+    }
+
+  }
+
+  {// Test when the element being searched is the last one
+
+    {
+      int a[] = {1, 2, 3, 4};
+      std::same_as<std::ranges::subrange<It>> auto ret = std::ranges::find_last_if(It(a), Sent(It(a + 4)), [](int x) { return x == 4; });
+      assert(base(ret.begin()) == a + 3);
+      assert(*ret.begin() == 4);
+    }
+
+    {
+      int a[] = {1, 2, 3, 4};
+      std::same_as<std::ranges::borrowed_subrange_t<int (&)[4]>> auto ret = std::ranges::find_last_if(a, [](int x) { return x == 4; });
+      assert(base(ret.begin()) == a + 3);
+      assert(*ret.begin() == 4);
+    }
+
+  }
+
+
+  {// check that past-the-end iterator is returned with no match
+
+    {
+      int a[] = {1, 1, 1};
+      std::same_as<std::ranges::subrange<int*>> auto ret = std::ranges::find_last_if(a, a + 3, [](int) { return false; });
+      assert(ret.data() == a + 3);
+    }
+
+    {
+      int a[]  = {1, 1, 1};
+      std::same_as<std::ranges::subrange<int*>> auto ret = std::ranges::find_last_if(a, [](int) { return false; });
+      assert(ret.data() == a + 3);
+    }
+
+  }
+
+}
+
+
+struct NonConstComparableValue {
+  friend constexpr bool operator==(const NonConstComparableValue&, const NonConstComparableValue&) { return false; }
+  friend constexpr bool operator==(NonConstComparableValue&, NonConstComparableValue&) { return false; }
+  friend constexpr bool operator==(const NonConstComparableValue&, NonConstComparableValue&) { return false; }
+  friend constexpr bool operator==(NonConstComparableValue&, const NonConstComparableValue&) { return true; }
+};
+
+constexpr bool test() {
+  test_iterators<const int*>();
+  test_iterators<int*>();
+  test_iterators<bidirectional_iterator<int*>>();
+  test_iterators<forward_iterator<int*>>();
+  test_iterators<random_access_iterator<int*>>();
+  test_iterators<contiguous_iterator<int*>>();
```

Comment:
Please have tests when sentinel is a different type than the iterator.
This applies for other tests